### ISSUE TITLE
reproduced result is low on ResNet backbone

### ISSUE THREAD

**Issue by youngwanLEE** [jfzhang95/pytorch-deeplab-xception#84]
I tried to reproduce the result(78.43%) of ResNet backbone in README.MD just by using [this command](https://github.com/jfzhang95/pytorch-deeplab-xception/blob/master/train_voc.sh).

Of course, I prepared for SBD dataset and train the model on VOC2012 + SBD. 

Then, my result was  75.06 is lower than the reported one.  How can I  get the reported result?

By the way, weird things were shown in the Tensorboard. while train loss converged, validation loss increased. 

![image](https://user-images.githubusercontent.com/13238578/57116235-ce927880-6d8e-11e9-91ea-c13d73ccef98.png)

**Comment by theevann** [comment]
I got the same curves - Did you find an explanation ? May the pytorch version make a difference ?

**Comment by cxxgtxy** [comment]
I also obtained a lower result. MIOU 74.58%

**Comment by lightas** [comment]
I obtained a similar low result too. mIOU 74.15%. Any solution?

**Comment by theevann** [comment]
@lightas I contacted the owner and he said that he froze batch norm during training.  
Be careful however that the --freeze-bn parameter **doesn't work** currently in the code !

Since I am interested in **Cityscapes**, I focused my time on it:

- I tried with frozen batch-norm (w/ cityscapes) but it didn't help.
- I obtained some correct results (74%) (w/ cityscapes) by lowering the LR at a faster rate using the step scheduler. So maybe you could try the step scheduler ?
- I am currently training again (w/ cityscapes) with the poly scheduler but changing the `lr_power` (in file `utils/lr_scheduler.py`): 
   ```python
    self.lr * pow((1 - 1.0 * T / self.N), lr_power)
    ```
   I'm trying `lr_power=3` and it is giving better results in my case, by lowering the learning rate faster.

I haven't retried on Pascal VOC yet, so I don't know if this makes a difference.

Also note that the new Torchvision v0.3 has deeplabv3 "built-in".

**Comment by lightas** [comment]
@theevann  Thank you so much. I will try it.

**Comment by lightas** [comment]
@theevann Hi, would you please tell me why you said that the --freeze-bn parameter doesn't work? I didn't find out why it doesn't work.

**Comment by theevann** [comment]
@lightas 
The freeze-bn parameter puts BatchNorn into eval mode at model initialization.
But then in the training you do model.train(), which sets the BatchNorm back to training mode...

**Comment by pengwangucla** [comment]
> he froze batch norm during training.

So the synchronized batch norm does not work in the code? Since sync bn should improve the performance, rather than decrease it

**Comment by lightas** [comment]
@theevann I got it. Thank you so much.

**Comment by beizhengren** [comment]
@youngwanLEE Hi, how to generate those curves? Thank you!

**Comment by yangninghua** [comment]
```
    def training(self, epoch):
        train_loss = 0.0
        self.model.train()
        for m in self.model.modules():
            if isinstance(m, SynchronizedBatchNorm2d):
                m.eval()
            elif isinstance(m, nn.BatchNorm2d):
                m.eval()
        tbar = tqdm(self.train_loader)
```

**Comment by yangninghua** [comment]
```
def make_data_loader(args, **kwargs):

    if args.dataset == 'pascal':
        train_set = pascal.VOCSegmentation(args, split='train')
        val_set = pascal.VOCSegmentation(args, split='val')
        if args.use_sbd:
            sbd_train = sbd.SBDSegmentation(args, split=['train', 'val'])
            train_set = combine_dbs.CombineDBs([train_set, sbd_train], excluded=[val_set])

        num_class = train_set.NUM_CLASSES
        num_class = 1
```

**Comment by wtsitp** [comment]
@youngwanLEE , what is you  training parameter, 
  the max mIoU is 0.61, I'm confused, thank you

**Comment by TRON15** [comment]
> @lightas I contacted the owner and he said that he froze batch norm during training.
> Be careful however that the --freeze-bn parameter **doesn't work** currently in the code !
> 
> Since I am interested in **Cityscapes**, I focused my time on it:
> 
> * I tried with frozen batch-norm (w/ cityscapes) but it didn't help.
> * I obtained some correct results (74%) (w/ cityscapes) by lowering the LR at a faster rate using the step scheduler. So maybe you could try the step scheduler ?
> * I am currently training again (w/ cityscapes) with the poly scheduler but changing the `lr_power` (in file `utils/lr_scheduler.py`):
>   ```python
>    self.lr * pow((1 - 1.0 * T / self.N), lr_power)
>   ```
>   
>   
>   I'm trying `lr_power=3` and it is giving better results in my case, by lowering the learning rate faster.
> 
> Note :
> 
> * Train on multiple GPUs, since a large enough batch-size is very important. (I used 4 GPU with bs 16 : 4 on each gpu).
> * Train with a small crop size : this allows to process more images per batch and stabilize the training. (I used 550x550)
> * Remove image-level features in the ASPP module of DeepLab. (Not a big deal though)
> 
> I haven't retried on Pascal VOC yet, so I don't know if this makes a difference.
> 
> Also note that the new Torchvision v0.3 has deeplabv3 "built-in".

@theevann Thanks a lot for sharing the useful modification. And I find the DeepLabv3+ paper didn't give the mIOU in on the resnet101 backbone, only Xception in the paper. And for cityscape, the [official repo's zoo ](https://github.com/tensorflow/models/blob/master/research/deeplab/g3doc/model_zoo.md) only add Xception and MobileNetv2 backbone. So how do you get the right mIOU range for DeepLabv3+ with Resnet101 as the backbone?
Also, could you tell us some more detail info about your experiments in CityScapes? Did you use the additional coarse data? And did you set the output stride =8 or 16?

Thanks a lot!

**Comment by GuoleiSun** [comment]
Same problem. I got 74.**% training on VOC2012 + SBD, and using Resnet. I can not get 78%. Seems that this code base is not good for reproduction.

**Comment by wtsitp** [comment]
what is your train config，
the acc is 61%，I train


| |
王涛
邮箱：wtsitp@163.com
|

签名由 网易邮箱大师 定制

On 10/28/2019 10:39, Guolei Sun wrote:

Same problem. I got 74.**% training on VOC2012 + SBD, and using Resnet. I can not get 78%. Seems that this code base is not good for reproduction.

**Comment by GuoleiSun** [comment]
I use "train_voc.py". Actually, I think my problem is exactly as the author of this issue. But the issue seems not solved.

**Comment by GuoleiSun** [comment]
Basically, what I want to do is to reproduce "78.43%" using "train_voc.py". Of course, I prepared for SBD dataset and train the model on VOC2012 + SBD. But I only got "74%", which is much lower than what I expect

**Comment by TRON15** [comment]
Hi guys, I will share some of my experiment settings.

Dataset: Cityscape without coarse additional data
Backbone: Resnet101
output_stride: 16
initial_learning_rate: 0.005
learning_decay: ploy
training_epochs: 240
batch_size: 8
num_gpus: 4
train_size: 768*768
**no sync_bn**
others: the default value of the train.py

Inference in 2048*1024 in val dataset of cityscapes, I got ~74% mIOU. I am not sure if it's a right mIOU

**Comment by GuoleiSun** [comment]
Hi guys,

For those who want to reproduce results on deeplab V3, I recommend this code: https://github.com/chenxi116/DeepLabv3.pytorch
The code can simply reproduce 76.8% mIOU in Pascal val (trained on VOC2012 + SBD).

**Comment by XUYUNYUN666** [comment]
> I also obtained a lower result. MIOU 74.58%

can I have your qq, i have a so poor performace on my owndataset and  trained via 8 gpus, thank u very much .

**Comment by XUYUNYUN666** [comment]
> Hi guys, I will share some of my experiment settings.
> 
> Dataset: Cityscape without coarse additional data
> Backbone: Resnet101
> output_stride: 16
> initial_learning_rate: 0.005
> learning_decay: ploy
> training_epochs: 240
> batch_size: 8
> num_gpus: 4
> train_size: 768*768
> **no sync_bn**
> others: the default value of the train.py
> 
> Inference in 2048*1024 in val dataset of cityscapes, I got ~74% mIOU. I am not sure if it's a right mIOU

Can I have your qq number, I really want to get your help. I trained on my dataset with 8 gpus and got so poor performance..
